File: 5_Systemidentifikation/MATLAB/testsignals.m
clear all

nSamples = 5000;

% create signals
% gaussian (normal) distribution 
gaussian = randn(nSamples, 1);
% uniform distribution
uniform = 2*rand(nSamples, 1) - 1;
% random binary sequence 
rbs = sign(gaussian);

figure();
subplot(3,1,1)
plot(gaussian);
title('gaussian')
subplot(3,1,2)
plot(uniform)
title('uniform')
ylim([-2 2])
subplot(3,1,3)
plot(rbs)
title('RBS')
ylim([-2 2])
xlabel('Index {\itn}')
grid on


% plot spectra
normFreq = linspace(0, 2, nSamples);
figure();
plot(normFreq, abs(fft(gaussian)));
hold on
plot(normFreq, abs(fft(uniform)));
plot(normFreq, abs(fft(rbs)));
xlim([0 1])
xlabel('Normalized frequency \Omega')
legend('gaussian', 'uniform', 'RBS')
grid on


% plot auto correlation
lagSamples = (-(nSamples-1) : nSamples-1);
figure();
plot(lagSamples, xcorr(gaussian) ./ max(xcorr(gaussian)));
hold on
plot(lagSamples, xcorr(uniform) ./ max(xcorr(uniform)));
plot(lagSamples, xcorr(rbs) ./ max(xcorr(rbs)));
legend('gaussian', 'uniform', 'RBS')
xlabel('Index lag \Delta{\itn}')
ylabel('Normalized autocorrelation')
grid on

File: 5_Systemidentifikation/Python/testsignals.py
import numpy as np
import matplotlib.pyplot as pp

n_samples = 5000

# create signals
# gaussian (normal) distribution
gaussian = np.random.randn(n_samples)
# uniform distribution
uniform = 2*np.random.rand(n_samples) - 1
# random binary sequence
rbs = np.sign(gaussian)

pp.figure()
pp.subplot(311)
pp.plot(gaussian)
pp.title('gaussian')
pp.subplot(312)
pp.plot(uniform)
pp.title('uniform')
pp.ylim(-2, 2)
pp.subplot(313)
pp.plot(rbs)
pp.title('RBS')
pp.ylim(-2, 2)
pp.xlabel('Index $n$')
pp.grid(True)


# plot spectra
norm_freq = np.linspace(0, 2, n_samples)
pp.figure()
pp.plot(norm_freq, abs(np.fft.fft(gaussian)))
pp.plot(norm_freq, abs(np.fft.fft(uniform)))
pp.plot(norm_freq, abs(np.fft.fft(rbs)))
pp.xlim(0, 1)
pp.xlabel('Normalized frequency $\Omega$')
pp.legend(('gaussian', 'uniform', 'RBS'))
pp.grid(True)


# plot auto correlation
lag_samples = np.concatenate((np.linspace(-(n_samples - 1), 0, n_samples - 1, endpoint=False),
                              np.linspace(0, n_samples, n_samples, endpoint=False)))
pp.figure()
pp.plot(lag_samples, np.correlate(gaussian, gaussian, 'full') /
        max(np.correlate(gaussian, gaussian, 'full')))
pp.plot(lag_samples, np.correlate(uniform, uniform, 'full') /
        max(np.correlate(uniform, uniform, 'full')))
pp.plot(lag_samples, np.correlate(rbs, rbs, 'full') /
        max(np.correlate(rbs, rbs, 'full')))
pp.legend(('gaussian', 'uniform', 'RBS'))
pp.xlabel('Index lag $\Delta n$')
pp.ylabel('Normalized autocorrelation')
pp.grid(True)

pp.show()
